url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/model/Task.java

```diff
@@ -0,0 +1,622 @@
+		return earliestSubTask;
+	}
+
+	/**
+	 * Get the sub-task, who has the earliest end time, in the sub-task tree include myself.
+	 * 
+	 * @return
+	 */
+	public Task getEarliestTask() {
+		Task earliestSubTask = this.getEarlistOrLatestSubTask(this, true);
+		if (earliestSubTask != null && earliestSubTask.getActualStart().after(this.getActualStart())) {
+			earliestSubTask = this;
+		}
+		return earliestSubTask;
+	}
+
+	/**
+	 * Get the last task, who has the latest end time, in the sub task tree.
+	 * 
+	 * @return
+	 */
+	public Task getLatestSubTask() {
+		Task latestSubTask = this.getEarlistOrLatestSubTask(this, false);
+		return latestSubTask;
+	}
+
+	/**
+	 * Get the last task, who has the latest end time, in the sub task tree include myself.
+	 * 
+	 * @return
+	 */
+	public Task getLatestTask() {
+		Task latestSubTask = this.getEarlistOrLatestSubTask(this, false);
+		// Can not include myself because the root task node should be excluded. 
+		if (latestSubTask != null && latestSubTask.getActualEnd().before(this.getActualEnd())) {
+			latestSubTask = this;
+		}
+		return latestSubTask;
+	}
+
+	/*
+	 * Get latest sub-task in the tree recursively, excludes myself.
+	 */
+	private Task getEarlistOrLatestSubTask(Task root, boolean earlist) {
+		Task ret = null;
+		for (int i = 0; i < root.getChildCount(); i++) {
+			Task curChild = (Task) root.children.get(i);
+			Task current = null; // Current for earliest or latest in sub-tree.
+			if (curChild.isLeaf() == false) {// Get latest child if has children.
+				current = getEarlistOrLatestSubTask(curChild, earlist);
+			}
+			if (earlist == true) {// Get earliest sub  task.
+				if (current == null || current.getActualStart().after(curChild.getActualStart())) {// Compare to current child.
+					current = curChild;
+				}
+				//Update the indicator if earlier task found in current sub-tree. 
+				if (ret == null || ret.getActualStart().after(current.getActualStart())) {
+					ret = current;
+				}
+			} else { // Get latest sub task.
+				if (current == null || current.getActualEnd().before(curChild.getActualEnd())) {// Compare to current child.
+					current = curChild;
+				}
+				//Update the indicator if later task found in current sub-tree. 
+				if (ret == null || ret.getActualEnd().before(current.getActualEnd())) {
+					ret = current;
+				}
+			}
+		}
+		return ret;
+	}
+
+	/**
+	 * Calculate how many steps for this task in Gantt chart.
+	 * 
+	 * @param task
+	 * @return
+	 */
+	public int calcTaskSteps() {
+		if (this.treeModel == null || this.treeModel.getTimeUnit() == null) {
+			return 0;
+		}
+		TimeUnit timeUnit = this.treeModel.getTimeUnit();
+		int steps = TimeAxisUtils.getTimeIntervalByTimeUnit(timeUnit, this.getActualStart(), this.getActualEnd()) + 1;
+		return steps;
+	}
+
+	/**
+	 * Calculate displaying progress steps by task progress.
+	 * 
+	 * @param start
+	 * @param progress
+	 * @return
+	 */
+	public int calcProgressSteps() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.